sycl/include/sycl/ext/intel/experimental/esimd/math.hpp

```diff
@@ -1371,21 +1371,35 @@ template <> ESIMD_INLINE float atan2(float y, float x) {
 template <int N>
 ESIMD_INLINE __ESIMD_NS::simd<float, N> fmod(__ESIMD_NS::simd<float, N> y,
                                              __ESIMD_NS::simd<float, N> x) {
-  __ESIMD_NS::simd<int, N> v_quot;
   __ESIMD_NS::simd<float, N> fmod;
+  __ESIMD_NS::simd<float, N> abs_x;
+  __ESIMD_NS::simd<float, N> reminder;
+  __ESIMD_NS::simd<float, N> reminder_sign_mask;
+  __ESIMD_NS::simd<float, N> y_sign_mask;
 
-  v_quot = convert<int>(y / x);
-  fmod = y - x * convert<float>(v_quot);
-  return fmod;
+  y_sign_mask.merge(-1.0f, 1.0f, y < 0);
+  abs_x = __ESIMD_NS::abs(x);
+  reminder = y - abs_x * __ESIMD_NS::trunc<float>(__ESIMD_NS::abs(y) / abs_x);
+  reminder_sign_mask.merge(1.0f, 0.0f, reminder < 0);
+
+  fmod = reminder + abs_x * reminder_sign_mask;
+  return __ESIMD_NS::abs(fmod) * y_sign_mask;
```

Comment:
This implementation is better and more correct, but still not fully compatible with std::fmod.

std::fmod(-0.0f, 1.0f) would return -0.0f, while the implementation here returns "+0.0f".
The LIT test (https://github.com/intel/llvm-test-suite/pull/1045) passes because it compares the expected and computed results using '==', returning true even though they are not bitwise equal.

@akolesov-intel can you please give some advice on how to implement it the most efficient way that would be also follow std::fmod rules?

Reply:
Hi Slava, you are correct. Fully compliant fmod implementation is not that simple. See attached codes for single and double precision cases. As usual there are fast paths (__imf_fmodf, __imf_fmod) and slow ones for large and special arguments (__imf_internal_sfmod, __imf_internal_dfmod). These implementations are fully compliant to standard and return exact results
[fmod.zip](https://github.com/intel/llvm/files/8911214/fmod.zip)
.